### PR TITLE
InnoDB migration solution 2

### PR DESCRIPTION
**Pull request by matrix-amiel**
Here's new branch with InnoDB migration for all OpenEMR  tables.
For the 4 MyISAM tables (with auto_increment field in the end of primary key) the solution presented in this branch is to calculate a new value for former auto_increment fields inside  insert query . (What was called ‘Solution 2’ here  http://www.open-emr.org/wiki/index.php/InnoDB_Transition) 

This code deprecates branch OS/InnoDB and pull request 124.

@matrix-olegsv  @bradymiller @sunsetsystems

### CODE REVIEW

File: library/sql.inc [hunk 1]
```diff
@@ -650,4 +650,31 @@ function generic_sql_insert_id()
...
+function sqlBeginTrans()
+{
+    $GLOBALS['adodb']['db']->BeginTrans();
+}
+
+
+/**
+ * Commit a transaction
+ */
+function sqlCommitTrans($ok=true)
```
Comment (bradymiller) [inline library/sql.inc]: What is the $ok parameter used for?

File: sql/4_2_2-to-4_3_1_upgrade.sql [hunk 2]
```diff
@@ -71,4 +71,129 @@
...
+
+#IfTableEngine procedure_order MyISAM
+-- remove NOT NULL DEFAULT "" declaration from TEXT field.
+ALTER TABLE `procedure_order` MODIFY `patient_instructions` TEXT, ENGINE=InnoDB;
+#EndIf
+
+-- #IfTableEngine procedure_order_code MyISAM - see below
+-- remove NOT NULL DEFAULT "" declaration from TEXT field.
+-- ALTER TABLE `procedure_order_code` MODIFY `diagnoses` TEXT COMMENT 'diagnoses and maybe other coding (e.g. ICD9:111.11)', ENGINE=InnoDB;
+-- #EndIf
```
Comment (bradymiller) [inline sql/4_2_2-to-4_3_1_upgrade.sql]: Why is this block commented out?

File: sql/database.sql [hunk 3]
```diff
@@ -2662,9 +2662,9 @@ DROP TABLE IF EXISTS `lang_custom`;
 CREATE TABLE `lang_custom` (
   `lang_description` varchar(100) NOT NULL default '',
   `lang_code` char(2) NOT NULL default '',
-  `constant_name` mediumtext NOT NULL default '',
-  `definition` mediumtext NOT NULL default ''
-) ENGINE=MyISAM ;
+  `constant_name` mediumtext NOT NULL,
+  `definition` mediumtext
```
Comment (bradymiller) [inline sql/database.sql]: Just to get some clarification here for future coding.
1. "NOT NULL" is ok
2. Why were the two above entries changes to 2 different settings (one with nothing and other set to NOT NULL)?

File: library/sql.inc (hunk 1 above)
Comment (olegsv) [inline library/sql.inc]: Its not used, but as sqlXXX functions map to AdoDB functions we made this be compatible with CommitTrans($ok) function ( See drivers/adodb-mysqli.inc.php ).

File: sql/4_2_2-to-4_3_1_upgrade.sql (hunk 2 above)
Comment (olegsv) [inline sql/4_2_2-to-4_3_1_upgrade.sql]: We grouped tables according to the treatments they need to receive. procedure_order_code table needs modification for  AUTO_INCREMENT problem and for TEXT DEFAULT problem. This section was left commented out to indicate that it's still member of TEXT DEFAULT.

File: sql/database.sql [hunk 4]
```diff
@@ -6493,12 +6493,12 @@ CREATE TABLE ar_session (
...
   KEY user_closed (user_id, closed),
   KEY deposit_date (deposit_date)
-) ENGINE=MyISAM;
+) ENGINE=InnoDB;
 
 CREATE TABLE ar_activity (
   pid            int(11)       NOT NULL,
   encounter      int(11)       NOT NULL,
-  sequence_no    int unsigned  NOT NULL AUTO_INCREMENT,
+  sequence_no    int unsigned  NOT NULL            COMMENT 'Ar_activity sequence_no, incremented in code',
```
Comment (bradymiller) [inline sql/database.sql]: strange spacing here

Comment (bradymiller) [inline sql/database.sql]: spacing ok, disregard prior message

File: sql/database.sql (hunk 3 above)
Comment (olegsv) [inline sql/database.sql]: definition may not be empty null, but constant name may be ( in unfortunate situation )...

**Comment by olegsv** [comment]
@sunsetsystems this is not exactly solution 2 as we discussed. Due to a slight mismatch, we got here another version.  This is a transaction-based solution, not documented in Wiki. Do you want us to correct this or you'll do it youself?

**Comment by sunsetsystems** [comment]
@olegsv The method looks fine but I would still like to see it implemented with functions.  I'm happy to do that after your commit is done.  Thanks!

**Comment by olegsv** [comment]
@sunsetsystems  Then the branch / commit is yours.

**Comment by sunsetsystems** [comment]
Separate comments for separate contributions please.  :)

**Comment by olegsv** [comment]
I'm not sure I understand correctly. If you want us to completely redo the commits?

**Comment by sunsetsystems** [comment]
No, just saying I will do a separate commit after this one is in SourceForge.

File: sql/database.sql (hunk 3 above)
Comment (bradymiller) [inline sql/database.sql]: Hi Oleg,
So, to clarify, both "mediumtext" and "mediumtext NOT NULL" are ok in INNODB.
And then to clarify:
1. "mediumtext" will set to NULL if nothing is given?
2. "mediumtext NOT NULL" will set to ''(ie. blank) if nothing is given?

Comment (olegsv) [inline sql/database.sql]: MEDIUMTEXT NOT NULL will fail if nothing is given.
Judging from language.inc.php:36 
`     $sql="INSERT INTO lang_custom SET lang_code=?, lang_description=?";`

Can you please  change `constant_name` definition to MEDIUMTEXT ?

Comment (bradymiller) [inline sql/database.sql]: So, we basically can not set TEXT fields to default?
So, they will always then be NULL if not set?

Comment (bradymiller) [inline sql/database.sql]: If above is the case, shouldn't all the 'TEXT NOT NULL' items be changed?

Comment (olegsv) [inline sql/database.sql]: MySQL 5 does not allow to set default value to fields of certain types.
http://dev.mysql.com/doc/refman/5.7/en/data-type-defaults.html

It generates a warning and omits DEFAULT clause, so NOT NULL table becomes a null.

Frankly speaking, it is a bug not related to InnoDB.

I think that all TEXT NOT NULL items should be changed , but this change requires code inspection.

**Comment by bradymiller** [comment]
Did some housekeeping and created a new pull request to keep this code manageable:
1. Flattened all commits into 1 commit.
2. Based the commit on most recent codebase
Here is the new pull request:
https://github.com/openemr/openemr/pull/130

**Pull request closed without merging**